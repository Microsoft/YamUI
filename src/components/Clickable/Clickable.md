### Notes for use

Clickable should be used in place of NavigationLink when you need click interaction without navigation. By default a Clickable is styled like a link, but you can also set it to `unstyled`.


### Examples

Inline:

```js { "props": { "data-description": "inline", "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-clickable\"}]" } }
<div>
  The following Clickable example looks like a link but is actually rendered as a button: <Clickable onClick={action('clickable clicked')}>Click me</Clickable>
</div>
```

Unstyled:

```js { "props": { "data-description": "unstyled", "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-clickable\"}]" } }
<div>
  The following Clickable example has <strong>unstyled: true</strong>: <Clickable onClick={action('clickable clicked')} unstyled={true}>Click me</Clickable>
</div>
```

Unstyled block with FakeLink:

```js { "props": { "data-description": "unstyled block with fakelink", "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-clickable\"}]" } }
<div>
  <Clickable onClick={action('clickable clicked')} unstyled={true} block={true}>
    This entire area is a Clickable with <strong>unstyled: true</strong> and <strong>block: true</strong>. Here is a nested <FakeLink>FakeLink component</FakeLink>, which looks like a link and shows link hoverstate when any part of the Clickable is hovered. This can work with <FakeLink>multiple nested FakeLinks</FakeLink>.
  </Clickable>
</div>
```

Wrapping block content:

```js { "props": { "data-description": "wrapping block", "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-clickable\"}]" } }

<Clickable onClick={action('clickable clicked')} block={true}>
  <div>
    This <strong>Clickable</strong> component wraps an entire block of text. By default all text content within a <strong>Clickable</strong> tag will be styled like a link.
  </div>
</Clickable>
```

Wrapping image:

```js { "props": { "data-description": "wrapping image", "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-clickable\"}]" } }
const yammerLogo = 'logo.png';
const yammerLogoDescription = 'Yammer "y" logo';

<div>
  <Clickable onClick={action('clickable clicked')} unstyled={true} block={true}>
    <Image source={yammerLogo} description={yammerLogoDescription} width={320} height={240} block={true}/>
  </Clickable>
</div>
```
