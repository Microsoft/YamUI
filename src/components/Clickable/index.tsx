/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
import './clickable.css';

export interface ClickableProps extends NestableBaseComponentProps {
  block?: boolean;
  ariaLabel?: string;
  title?: string;
  onClick?: (event: any) => void;
}

const getClasses = (props: ClickableProps) => {
  const classes: string[] = ['y-clickable'];
  if (props.block) {
    classes.push('y-clickable__block');
  }
  if (props.className) {
    classes.push(props.className);
  }
  return classes.join(' ');
};

const Clickable: React.StatelessComponent<ClickableProps> = props => (
  <button
    className={getClasses(props)}
    onClick={props.onClick}
    aria-label={props.ariaLabel}
    title={props.title}
  >
    {props.children}
  </button>
);

export default Clickable;
