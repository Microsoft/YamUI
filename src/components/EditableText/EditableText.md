### Notes for use
EditableText will inherit the text size of the nearest Block component.

### Examples

With text and max length:

```js { "props": { "data-description": "with text and maxLength", "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-clickable\"},{\"action\":\"hover\",\"selector\":\".y-clickable\"}]" } }
const { TextSize } = require('../Block');

<Block textSize={TextSize.SMALL}>
  <EditableText text="A spreadsheet of business requirements" maxLength={40} />
</Block>
```

With prompt text and placeholder text:

```js { "props": { "data-description": "with promptText and placeHolder", "data-action-states": "[{\"action\":\"none\"},{\"action\":\"hover\",\"selector\":\".y-clickable\"},{\"action\":\"click\",\"selector\":\".y-clickable\"},{\"action\":\"keyPress\",\"key\":\"Y\"}]" } }
const { TextSize } = require('../Block');

<Block textSize={TextSize.SMALL}>
  <EditableText promptText="Click to edit..." placeHolder="Attachment description" />
</Block>
```

With onBeginEditing, onEndEditing and onUpdate callbacks:

```js { "props": { "data-description": "with callbacks" } }
const { TextSize } = require('../Block');

<Block textSize={TextSize.SMALL}>
  <EditableText
    text="Document description"
    onUpdate={action('onUpdate')}
    onBeginEditing={action('onBeginEditing')}
    onEndEditing={action('onEndEditing')}
  />
</Block>
```
