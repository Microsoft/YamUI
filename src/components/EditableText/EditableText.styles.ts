/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { mergeStyleSets, ITheme } from 'office-ui-fabric-react/lib/Styling';
import { memoizeFunction } from 'office-ui-fabric-react/lib/Utilities';

export interface EditableTextStyleProps {
  isEditing: boolean;
  theme: ITheme;
}

export const getClassNames = memoizeFunction((styleProps: EditableTextStyleProps) => {
  const { isEditing, theme } = styleProps;
  const activeColorStyle = { color: theme.semanticColors.buttonText };
  const underlineStyle = { textDecoration: 'underline' };

  const root = isEditing
    ? {}
    : {
        color: theme.semanticColors.bodySubtext,
        selectors: {
          ':hover': activeColorStyle,
          ':active': activeColorStyle,
          ':focus': activeColorStyle,
        },
      };

  return mergeStyleSets({
    root,
    clickableText: {
      paddingLeft: '0.2rem',
      selectors: {
        ':hover': underlineStyle,
        ':active': underlineStyle,
        ':focus': underlineStyle,
      },
    },
  });
});
