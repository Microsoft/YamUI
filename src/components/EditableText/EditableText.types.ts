/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import { BaseComponentProps } from '../../util/BaseComponent/props';

export interface EditableTextProps extends BaseComponentProps {
  /**
   * Current text.
   */
  text?: string;

  /**
   * The text to display when text is empty.
   */
  promptText?: string;

  /**
   * The TextField placeHolder text.
   */
  placeHolder?: string;

  /**
   * An optional max length for the description field when editing.
   */
  maxLength?: number;

  /**
   * Returns the new text string when updated. Triggered by enter keypress or blur event to close the textfield.
   */
  onUpdate?: ((text: string) => void);

  /**
   * Triggered when the user enters edit mode.
   */
  onBeginEditing?: (() => void);

  /**
   * Triggered when the user exits edit mode.
   */
  onEndEditing?: (() => void);
}

export interface EditableTextState {
  isEditing: boolean;
  editedValue: string;
}
