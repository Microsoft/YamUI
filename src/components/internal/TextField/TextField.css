/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
@import 'src/css/variables/colors.css';
@import 'src/css/variables/fonts.css';

/* work is in progress to set styling/theme options for all Fabric label instances */
.y-textField .ms-Label {
  @extend text__mediumSub;
  font-weight: $fontWeight__bold;
  color: $textColor__secondary;
  padding-bottom: 0.4rem;
}

.y-textField .ms-TextField-fieldGroup {
  line-height: 3.3rem;
  border-radius: 1px;
}

.y-textInput .ms-TextField-fieldGroup {
  min-height: 3.6rem;
}

.y-textField .ms-TextField-field {
  @extend text__mediumSub;
  color: $textColor;
  padding: 0 0.7rem;
}

.y-textField.is-disabled .ms-TextField-field {
  color: $textColor__disabled;
}

.y-textField--description,
.y-textField .ms-TextField-errorMessage {
  @extend text__small;
  display: block;
  padding-top: 0.2rem;
}

.y-textField--description {
  color: $textColor__metadata;
}

.y-textField .ms-TextField-errorMessage {
  color: $textColor__error;
}

/* TextArea styling */
.y-textField.ms-TextField--multiline .ms-TextField-field {
  @extend text__mediumSub;
  padding-top: 0.7rem;
  padding-bottom: 0.3rem;
}

/* TextInput styling */
.y-textField.ms-TextField--underlined .ms-TextField-fieldGroup {
  height: 2.5rem;
  padding-bottom: 0.3rem;
  min-height: initial;
}

.y-textField.ms-TextField--underlined .ms-TextField-field {
  padding-left: 0;
}

.y-textField.ms-TextField--underlined .y-textField--description,
.y-textField.ms-TextField--underlined .ms-TextField-errorMessage {
  padding-top: 0.1rem;
}

.y-textField.ms-TextField--underlined.is-active .ms-TextField-wrapper {
  border-color: $borderColor__focused;
}

.y-textField__with-prefix.ms-TextField--underlined .ms-TextField-field,
.y-textField__with-prefix .ms-TextField-field {
  padding-left: 0.4rem;
}

.y-textField--with-suffix.ms-TextField--underlined .ms-TextField-field,
.y-textField--with-suffix .ms-TextField-field {
  padding-right: 0.4rem;
}

.y-textField .ms-TextField-field {
  flex: 1;
  min-width: 30%;
}

.y-textField .ms-TextField-prefix,
.y-textField .ms-TextField-suffix {
  padding: 0 0.7rem 0.1rem 0.8rem;
  overflow: hidden;
}

.y-textField .ms-TextField-prefix span,
.y-textField .ms-TextField-suffix span {
  padding-right: 0.1rem;
  text-align: center;
}

.y-textField .ms-TextField-prefix span,
.y-textField .ms-TextField-suffix span {
  line-height: normal;
  overflow: hidden;
  text-overflow: ellipsis;
}
