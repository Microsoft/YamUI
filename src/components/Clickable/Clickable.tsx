/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import FakeLink from '../FakeLink';
import { camelCaseToDashed } from '../../util/classNames';
import { NestableBaseComponentProps } from '../../util/BaseComponent/props';
import './Clickable.css';

export interface ClickableProps extends NestableBaseComponentProps {
  /**
   * Additional label that must be provided if the clickable text is not descriptive enough.
   */
  ariaLabel?: string;

  /**
   * Whether the clickable should be `display: block`.
   */
  block?: boolean;

  /**
   * Display type to use. Defaults to inline, which allows contents to wrap across lines, but can't
   * hold block content. Use `Block` to contain block content, or `InlineBlock` to contain block
   * and size to block content
   */
  display?: 'inline' | 'block' | 'inlineBlock';

  /**
   * Focus outline style to use. Defaults to using 1px inset focus. If link contains content that
   * covers its surface, it should use displayType=BLOCK and focusType=OVERLAY so that the focus
   * is visible and not hidden under the children.
   */
  focus?: 'normal' | 'overlay' | 'highContrastInset';

  /**
   * Title or description of the linked document.
   */
  title?: string;

  /**
   * Whether to remove all styles from the link. Useful for allowing a large area to be clickable
   * while nesting `FakeLink` components to show link and hover state visuals.
   */
  unstyled?: boolean;

  /**
   * Click callback handler.
   */
  onClick?: ((event: React.MouseEvent<HTMLButtonElement>) => void);
}

/**
 * A `Clickable` is an accessible, clickable area that accepts arbitrary children. It is styled
 * like a link by default, but can also be unstyled. Under the hood `Clickable` simply wraps
 * content in a `button` element.
 */
export default class Clickable extends React.Component<ClickableProps> {
  public render() {
    const { ariaLabel, title, unstyled, onClick, children } = this.props;

    return (
      <button className={this.getClasses()} aria-label={ariaLabel} title={title} onClick={onClick} type="button">
        {unstyled ? children : <FakeLink>{children}</FakeLink>}
      </button>
    );
  }

  private getClasses() {
    const { className, display, focus } = this.props;

    const classes: string[] = ['y-clickable'];
    if (focus) {
      classes.push(`y-focus-${camelCaseToDashed(focus)}`);
    } else {
      classes.push('y-focus-normal');
    }

    if (display) {
      classes.push(`y-display-${camelCaseToDashed(display)}`);
    }

    if (className) {
      classes.push(className);
    }

    // TODO: only some combinations of block + focus are valid
    // TODO: block vs displayType

    return classes.join(' ');
  }
}
