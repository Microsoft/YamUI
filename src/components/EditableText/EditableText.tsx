/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import '../../yamui';
import * as React from 'react';
import { join } from '../../util/classNames';
import { BaseComponentProps } from '../../util/BaseComponent/props';
import Clickable from '../Clickable';
import EditIcon from '../Icon/icons/Edit';
import TextField, { ITextField } from '../TextField';
import { KeyCodes } from '../../util/enums';
import './EditableText.css';

export interface EditableTextProps extends BaseComponentProps {
  /**
   * Current text.
   */
  text?: string;

  /**
   * The text to display when text is empty.
   */
  promptText?: string;

  /**
   * The TextField placeHolder text.
   */
  placeHolder?: string;

  /**
   * An optional max length for the description field when editing.
   */
  maxLength?: number;

  /**
   * Returns the new text string when updated. Triggered by enter keypress or blur event to close the textfield.
   */
  onUpdate?: ((text: string) => void);

  /**
   * Triggered when the user enters edit mode.
   */
  onBeginEditing?: (() => void);

  /**
   * Triggered when the user exits edit mode.
   */
  onEndEditing?: (() => void);
}

export interface EditableTextState {
  isEditing: boolean;
  editedValue: string;
}

/**
 * Displays text which can be edited on click.
 */
export default class EditableText extends React.Component<EditableTextProps, EditableTextState> {
  private textFieldRef: ITextField | null;

  constructor(props: EditableTextProps) {
    super(props);
    this.textFieldRef = null;
    this.state = { isEditing: false, editedValue: '' };
  }

  public render() {
    const { editedValue, isEditing } = this.state;
    const { text, promptText, placeHolder, maxLength, className } = this.props;
    const classes = join(['y-editableText', className]);

    if (isEditing) {
      return (
        <div className={classes}>
          <TextField
            className="y-editableText--textfield"
            underlined={true}
            onChange={this.updateInternalEditedDescription}
            value={editedValue}
            placeHolder={placeHolder}
            maxLength={maxLength}
            componentRef={this.setTextFieldRef}
            onBlur={this.commitEdit}
            onKeyDown={this.onKeyDown}
          />
        </div>
      );
    }

    return (
      <span className={classes}>
        <Clickable
          onClick={this.handleEditClick}
          unstyled={true}
          className="y-editableText--clickable"
          ariaLabel={promptText}
        >
          <EditIcon /> <span className="y-editableText--clickableText">{text || promptText}</span>
        </Clickable>
      </span>
    );
  }

  public componentDidUpdate(_prevProps: EditableTextProps, prevState: EditableTextState) {
    // Set focus when we enter edit mode; better than depending on a hacky setTimeout()
    if (this.state.isEditing && !prevState.isEditing) {
      this.setTextFieldFocus();
    }
  }

  private onKeyDown = (event: any) => {
    switch (event.which) {
      case KeyCodes.escape:
        event.preventDefault();
        this.cancelEdit();
        break;
      case KeyCodes.enter:
        event.preventDefault();
        this.commitEdit();
        break;
      default:
    }
  };

  private setTextFieldRef = (component: ITextField | null) => {
    this.textFieldRef = component;
  };

  private updateInternalEditedDescription = (description: string) => {
    this.setState(() => ({ editedValue: description }));
  };

  private setTextFieldFocus() {
    if (this.textFieldRef) {
      this.textFieldRef.focus();
    }
  }

  private handleEditClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();
    this.setState(() => ({ isEditing: true, editedValue: this.props.text || '' }));
    if (this.props.onBeginEditing) {
      this.props.onBeginEditing();
    }
  };

  private cancelEdit = () => {
    this.setState(() => ({ isEditing: false, editedValue: '' }));
    if (this.props.onEndEditing) {
      this.props.onEndEditing();
    }
  };

  private commitEdit = () => {
    const { editedValue } = this.state;
    this.setState(() => ({ isEditing: false, editedValue: '' }));
    if (this.props.onUpdate) {
      this.props.onUpdate(editedValue);
    }
    if (this.props.onEndEditing) {
      this.props.onEndEditing();
    }
  };
}
