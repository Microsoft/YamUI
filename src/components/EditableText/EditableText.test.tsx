/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Clickable from '../Clickable';
import TextField, { TextFieldComponent } from '../TextField';
import EditableText, { EditableTextProps } from './index';

describe('<EditableText />', () => {
  let component: ShallowWrapper<EditableTextProps>;
  let fakeEvent: any;

  beforeEach(() => {
    fakeEvent = {
      preventDefault: jest.fn(),
      stopPropagation: jest.fn(),
    };
  });

  describe('with promptText', () => {
    beforeEach(() => {
      component = shallow(<EditableText promptText="PROMPT TEXT" />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with additional className', () => {
    beforeEach(() => {
      component = shallow(<EditableText className="TEST_CLASSNAME" />);
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('with all props except promptText', () => {
    let updateCallback: jest.Mock;
    let beginCallback: jest.Mock;
    let endCallback: jest.Mock;

    beforeEach(() => {
      updateCallback = jest.fn();
      beginCallback = jest.fn();
      endCallback = jest.fn();

      component = shallow(
        <EditableText
          text="TEXT"
          placeHolder="PLACEHOLDER"
          maxLength={120}
          onUpdate={updateCallback}
          onBeginEditing={beginCallback}
          onEndEditing={endCallback}
        />,
      );
    });

    it('matches its snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    describe('when clicked', () => {
      beforeEach(() => {
        component.find(Clickable).simulate('click', fakeEvent);
      });

      it('enters edit mode', () => {
        expect(component).toMatchSnapshot();
      });

      it('triggers the onBeginEditing callback', () => {
        expect(beginCallback).toHaveBeenCalledTimes(1);
      });

      describe('and text is changed', () => {
        beforeEach(() => {
          const onChange: (text: string) => void = component.find(TextField).prop('onChange') as (text: string) => void;
          onChange('new text');
        });

        describe('and ENTER is pressed', () => {
          beforeEach(() => {
            const onKeyDown: (event: any) => void = component.find(TextField).prop('onKeyDown') as (event: any) => void;
            onKeyDown({ which: 13, preventDefault: jest.fn() });
            component.update();
          });

          it('exits edit mode', () => {
            expect(component).toMatchSnapshot();
          });

          it('triggers the update callback with the correct value', () => {
            expect(updateCallback).toHaveBeenCalledWith('new text');
          });

          it('triggers the onEndEditing callback', () => {
            expect(endCallback).toHaveBeenCalledTimes(1);
          });
        });

        describe('and ESC is pressed', () => {
          beforeEach(() => {
            const onKeyDown: (event: any) => void = component.find(TextField).prop('onKeyDown') as (event: any) => void;
            onKeyDown({ which: 27, preventDefault: jest.fn() });
            component.update();
          });

          it('exits edit mode', () => {
            expect(component).toMatchSnapshot();
          });

          it('does not trigger the update callback', () => {
            expect(updateCallback).not.toHaveBeenCalled();
          });

          it('triggers the onEndEditing callback', () => {
            expect(endCallback).toHaveBeenCalledTimes(1);
          });
        });

        describe('and an unimportant key is pressed', () => {
          beforeEach(() => {
            const onKeyDown: (event: any) => void = component.find(TextField).prop('onKeyDown') as (event: any) => void;
            onKeyDown({ which: 65, preventDefault: jest.fn() });
            component.update();
          });

          it('remains in edit mode', () => {
            expect(component).toMatchSnapshot();
          });

          it('does not trigger the update callback', () => {
            expect(updateCallback).not.toHaveBeenCalled();
          });

          it('does not trigger the onEndEditing callback', () => {
            expect(endCallback).not.toHaveBeenCalled();
          });
        });

        describe('when the underlying TextField provides its ref', () => {
          let fakeRef: Partial<TextFieldComponent>;

          beforeEach(() => {
            // Get a reference to our fakeRef so we can check its focus() spy
            fakeRef = { focus: jest.fn() };
            const setRef: (node: Partial<TextFieldComponent>) => void = component
              .find(TextField)
              .prop('componentRef') as (node: Partial<TextFieldComponent>) => void;
            setRef(fakeRef);
          });

          describe('and the textfield enters edit mode', () => {
            beforeEach(() => {
              // Exit edit mode...
              const onKeyDown: (event: any) => void = component.find(TextField).prop('onKeyDown') as (
                event: any,
              ) => void;
              onKeyDown({ which: 27, preventDefault: jest.fn() });
              component.update();

              //  ...and re-enter edit mode to trigger focus now that we have a spy in place
              component.find(Clickable).simulate('click', fakeEvent);
            });

            it('focuses its text input', () => {
              expect(fakeRef.focus).toHaveBeenCalledTimes(1);
            });
          });
        });
      });
    });
  });

  describe('without props', () => {
    beforeEach(() => {
      component = shallow(<EditableText />);
    });

    describe('when clicked', () => {
      beforeEach(() => {
        component.find(Clickable).simulate('click', fakeEvent);
      });

      describe('and text is changed', () => {
        beforeEach(() => {
          const onChange: (text: string) => void = component.find(TextField).prop('onChange') as (text: string) => void;
          onChange('new text');
        });

        describe('and ENTER is pressed', () => {
          beforeEach(() => {
            const onKeyDown: (event: any) => void = component.find(TextField).prop('onKeyDown') as (event: any) => void;
            onKeyDown({ which: 13, preventDefault: jest.fn() });
            component.update();
          });

          it('exits edit mode', () => {
            expect(component).toMatchSnapshot();
          });
        });

        describe('and ESC is pressed', () => {
          beforeEach(() => {
            const onKeyDown: (event: any) => void = component.find(TextField).prop('onKeyDown') as (event: any) => void;
            onKeyDown({ which: 27, preventDefault: jest.fn() });
            component.update();
          });

          it('exits edit mode', () => {
            expect(component).toMatchSnapshot();
          });
        });

        describe('when the underlying TextField provides its ref', () => {
          let fakeRef: Partial<TextFieldComponent>;

          beforeEach(() => {
            // Get a reference to our fakeRef so we can check its focus() spy
            fakeRef = { focus: jest.fn() };
            const setRef: (node: Partial<TextFieldComponent>) => void = component
              .find(TextField)
              .prop('componentRef') as (node: Partial<TextFieldComponent>) => void;
            setRef(fakeRef);
          });

          describe('and the textfield enters edit mode', () => {
            beforeEach(() => {
              // Exit edit mode and re-enter edit mode to trigger focus now that we have a spy in place
              const onKeyDown: (event: any) => void = component.find(TextField).prop('onKeyDown') as (
                event: any,
              ) => void;
              onKeyDown({ which: 27, preventDefault: jest.fn() });
              component.update();

              component.find(Clickable).simulate('click', fakeEvent);
            });

            it('focuses its text input', () => {
              expect(fakeRef.focus).toHaveBeenCalledTimes(1);
            });
          });
        });
      });
    });
  });
});
