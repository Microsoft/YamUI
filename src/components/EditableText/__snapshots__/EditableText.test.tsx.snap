// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditableText /> with additional className matches its snapshot 1`] = `
<span
  className="y-editableText TEST_CLASSNAME"
>
  <Clickable
    className="y-editableText--clickable"
    focusableRef={[Function]}
    onClick={[Function]}
    unstyled={true}
  >
    <Edit />
    <span
      className="y-editableText--clickableText"
    />
  </Clickable>
</span>
`;

exports[`<EditableText /> with all props except promptText matches its snapshot 1`] = `
<span
  className="y-editableText"
>
  <Clickable
    className="y-editableText--clickable"
    focusableRef={[Function]}
    onClick={[Function]}
    unstyled={true}
  >
    <Edit />
    <span
      className="y-editableText--clickableText"
    >
      TEXT
    </span>
  </Clickable>
</span>
`;

exports[`<EditableText /> with all props except promptText when clicked and text is changed and ENTER is pressed exits edit mode 1`] = `
<span
  className="y-editableText"
>
  <Clickable
    className="y-editableText--clickable"
    focusableRef={[Function]}
    onClick={[Function]}
    unstyled={true}
  >
    <Edit />
    <span
      className="y-editableText--clickableText"
    >
      TEXT
    </span>
  </Clickable>
</span>
`;

exports[`<EditableText /> with all props except promptText when clicked and text is changed and ESC is pressed exits edit mode 1`] = `
<span
  className="y-editableText"
>
  <Clickable
    className="y-editableText--clickable"
    focusableRef={[Function]}
    onClick={[Function]}
    unstyled={true}
  >
    <Edit />
    <span
      className="y-editableText--clickableText"
    >
      TEXT
    </span>
  </Clickable>
</span>
`;

exports[`<EditableText /> with all props except promptText when clicked and text is changed and an unimportant key is pressed remains in edit mode 1`] = `
<div
  className="y-editableText"
>
  <TextField
    className="y-editableText--textfield"
    focusableRef={[Function]}
    maxLength={120}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeHolder="PLACEHOLDER"
    underlined={true}
    value="new text"
  />
</div>
`;

exports[`<EditableText /> with all props except promptText when clicked enters edit mode 1`] = `
<div
  className="y-editableText"
>
  <TextField
    className="y-editableText--textfield"
    focusableRef={[Function]}
    maxLength={120}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeHolder="PLACEHOLDER"
    underlined={true}
    value="TEXT"
  />
</div>
`;

exports[`<EditableText /> with all props except promptText when clicked when triggered to enter edit mode again remains in edit mode 1`] = `
<div
  className="y-editableText"
>
  <TextField
    className="y-editableText--textfield"
    focusableRef={[Function]}
    maxLength={120}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeHolder="PLACEHOLDER"
    underlined={true}
    value="TEXT"
  />
</div>
`;

exports[`<EditableText /> with all props except promptText when clicked when triggered to enter edit mode without private textFieldRef remains in edit mode 1`] = `
<div
  className="y-editableText"
>
  <TextField
    className="y-editableText--textfield"
    focusableRef={[Function]}
    maxLength={120}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeHolder="PLACEHOLDER"
    underlined={true}
    value="TEXT"
  />
</div>
`;

exports[`<EditableText /> with promptText matches its snapshot 1`] = `
<span
  className="y-editableText"
>
  <Clickable
    ariaLabel="PROMPT TEXT"
    className="y-editableText--clickable"
    focusableRef={[Function]}
    onClick={[Function]}
    unstyled={true}
  >
    <Edit />
    <span
      className="y-editableText--clickableText"
    >
      PROMPT TEXT
    </span>
  </Clickable>
</span>
`;

exports[`<EditableText /> without props when clicked and text is changed and ENTER is pressed exits edit mode 1`] = `
<span
  className="y-editableText"
>
  <Clickable
    className="y-editableText--clickable"
    focusableRef={[Function]}
    onClick={[Function]}
    unstyled={true}
  >
    <Edit />
    <span
      className="y-editableText--clickableText"
    />
  </Clickable>
</span>
`;

exports[`<EditableText /> without props when clicked and text is changed and ESC is pressed exits edit mode 1`] = `
<span
  className="y-editableText"
>
  <Clickable
    className="y-editableText--clickable"
    focusableRef={[Function]}
    onClick={[Function]}
    unstyled={true}
  >
    <Edit />
    <span
      className="y-editableText--clickableText"
    />
  </Clickable>
</span>
`;
